### PR TITLE
Refactor main function

### PR DESCRIPTION
**Pull request by THenry14**
Related to #86 

## Introduced changes

- This PR is a part of the stack that aims to allow for raw text profiling.

## Stack:
-- Set new interface (https://github.com/software-mansion/cairo-profiler/pull/130)
--> Clean up main function (This PR)
-- Add view (raw text profiling) logic (https://github.com/software-mansion/cairo-profiler/pull/132)
-- Add tests for new functionality (https://github.com/software-mansion/cairo-profiler/pull/133)
-- Add docs (https://github.com/software-mansion/cairo-profiler/pull/134)

### CODE REVIEW

File: crates/cairo-profiler/src/profile_builder.rs [hunk 1]
```diff
@@ -267,3 +276,27 @@ pub fn build_profile(samples: &[Sample]) -> pprof::Profile {
...
+        .expect("Failed to encode the profile to the buffer");
+
+    let mut buffer_reader = buffer.reader();
+    let mut encoder = GzEncoder::new(&mut buffer_reader, Compression::default());
+
+    let mut encoded_buffer = vec![];
+    encoder
+        .read_to_end(&mut encoded_buffer)
+        .context("Failed to read bytes from the encoder")?;
+    file.write_all(&encoded_buffer).unwrap();
```
Comment (ksew1) [inline crates/cairo-profiler/src/profile_builder.rs:299]: [super nit]
```suggestion
            fs::write(&build_cli.output_path, &encoded)?;
```

This way we don't need 

```
let mut file = fs::File::create(target_path).context("Failed to create the output file")?;
```

File: crates/cairo-profiler/src/profile_builder.rs [hunk 2]
```diff
@@ -267,3 +276,27 @@ pub fn build_profile(samples: &[Sample]) -> pprof::Profile {
...
+    if let Some(parent) = target_path.parent() {
+        fs::create_dir_all(parent)
+            .context("Failed to create parent directories for the output file")?;
+    }
+    let mut file = fs::File::create(target_path).context("Failed to create the output file")?;
+
+    let mut buffer = BytesMut::new();
+    profile
+        .encode(&mut buffer)
+        .expect("Failed to encode the profile to the buffer");
```
Comment (ksew1) [inline crates/cairo-profiler/src/profile_builder.rs:290]: Is there any guide lines when do we use here `expect` and when `result`? I some times find this inconsistent

```suggestion
        .context("Failed to encode the profile to the buffer")?;
```

**Review by ksew1 — APPROVED** [review]
LGTM!
200w

File: crates/cairo-profiler/src/profile_builder.rs (hunk 2 above)
Comment (THenry14) [reply crates/cairo-profiler/src/profile_builder.rs:290]: My rule of thumb is:
- is something unexpected? did you expect it should work? `.expect()`, cuz this is potentially the bug in our code, so we panic
- is something somewhat expected? possibly user induced in some way, or something the user has a control of? `.context()`, as an error (with additional context) suffices

But I know that this is pretty mixed in the profiler codebase. Maybe one day I take it upon myself and put it all in order :D

Comment (ksew1) [reply crates/cairo-profiler/src/profile_builder.rs:290]: Yep, I have a similar approach. I just wanted to ask so we are on the same page. I guess this is a valid case for using `expect`.

Maybe this should also be an `expect`, as we expect the reading buffer to pass:

```rust
encoder
.read_to_end(&mut encoded_buffer)
.context("Failed to read bytes from the encoder")?;
```

Nevertheless, it's such a minor thing that I'll probably stop making these kinds of comments, as they don't bring too much value to the PRs.

Comment (THenry14) [reply crates/cairo-profiler/src/profile_builder.rs:290]: thanks for pointing this out btw! ❤️

File: crates/cairo-profiler/src/profile_builder.rs [hunk 3]
```diff
@@ -267,3 +273,26 @@ pub fn build_profile(samples: &[Sample]) -> pprof::Profile {
         default_sample_type: 0,
     }
 }
+
+pub fn save_profile(target_path: &Utf8PathBuf, profile: &pprof::Profile) -> Result<()> {
```
Comment (piotmag769) [inline crates/cairo-profiler/src/profile_builder.rs:277]: I believe this function shouldn't be in this module as it is specifically designed for converting profiler's internal representation to pprof representation. Adding saving to a file here mixes abstraction layers imo and makes the purpose of the module less clear

**Pull request merged**